Rewrite indentation code

axvr opened the pull request:
> [!Tip]
> As of the latest change, indentation is 2–3x faster with the potential to surpass 10x with Vim9script and Lua.

This PR contains a rewrite of the entire Clojure indentation code that aims to:

1. Improve the (currently abysmal) indent performance and unlock yet more improvements in future,
   - #6
   - #32
   - #35
2. Change the defaults to be more similar to the [Clojure Style Guide](https://guide.clojure.style), [cljfmt](https://github.com/weavejester/cljfmt), [Emacs](https://github.com/clojure-emacs/clojure-mode) and [VS Code](https://calva.io/) defaults,
   - #22
3. Enable greater configuration of indent rules (such that various indentation styles are possible),
   - #21
5. Simplify the indentation code as much as possible (currently so complex, it is near impossible to work on),
6. Improve accuracy of the indentation algorithms, eliminating most (if not all) previous indentation bugs,
   - #34
   - https://github.com/clojure-vim/clojure.vim/pull/31#issuecomment-2227410501
7. Eliminate reliance on syntax highlighting to be compatible with alternate syntax highlighting systems (e.g. Tree-sitter),
   - #33
9. Create a rich indentation test suite,
   - #1
   - #39
   - #26
10. Make quality of life improvements (when the (Neo)Vim version supports them), such as preventing `=` operator from altering the indentation of multi-line strings and setting the `'lisp'` option in Clojure buffers.
    - #40


## To do

- [x] New indentation algorithm.
- [x] More indentation tests + test the different config options.
- [x] Run indentation tests on Vim and Neovim.
- [x] Update documentation (doc file and README).
- [x] Replace existing configuration options.
- [ ] Resolve most remaining TODOs.
- [ ] Create a GitHub issue for the remaining missing indent features:
  - Indentation rules for reader conditionals (i.e. `#?@(...)`, `#?(...)`).
  - Complex indentation rules (e.g. `letfn`, `extend-protocol`, etc.).
- [ ] Maybe write blog post about this update?

axvr committed:
Begin rewrite of indentation code

The objectives are:

    1. Simplify the indentation code; previous implementation has become
       so complex it is impossible to maintain,

    2. Significantly improve performance; previous indentation code was
       painfully slow, (see issue #6)

    3. Maximum configurability; should be configured similarly to cljfmt
       and make previously impossible things possible (e.g. issue #21).

As of this commit, objectives 1 and 2 have been met, but work on
objective 3 has not yet begun.  There will continue to be further
improvements, particularly around performance and the "what if syntax
highlighting is disabled?" scenario.

These changes will unfortunately be backwards incompatible, but
hopefully the improved performance and API will make up for it.

axvr committed:
Double (or better) indent performance!

By utilising some ugly mutable state, this change more than doubles the
performance of the indentation code.  You can expect even further
benefits in larger code blocks.  This should completely eliminate any
need for the `clojure_maxlines` configuration option.

axvr committed:
A bit of code clean up and minor optimisations

axvr committed:
Check for maps before vectors for further performance enhancement

Since maps are far more likely to across multiple lines than vectors
(and when they are they tend to be much larger, e.g. EDN files!), we can
further optimise indentation by checking if the line is inside a map
*before* checking it is inside a vector.

(This happens because of the performance improvement made with the
addition of the optimised `CheckPair` function.)

axvr committed:
Add fallback for when Vim was not compiled with `searchpairpos`

axvr committed:
Fix string detection when entering a newline while in insert mode

axvr committed:
Minor performance improvement when file contains multi-line strings

axvr committed:
Mimic multi-line string/regex indent behaviour of VS Code and Emacs

The `=` operator will no longer alter the indent of lines within
a Clojure multi-line string or regular expression.

The previous behaviour was annoying for writing detailed doc-strings, as
it made reformatting the file with `gg=G` not possible as it would screw
up the indentation within the doc-strings.

Now the behaviour matches that of VS Code and Emacs.

axvr committed:
When `clojure_align_multiline_strings` is `-1`, no indentation

By setting the `clojure_align_multiline_strings` option to `-1` it will
switch to traditional Lisp multi-line string indentation, of no indent.

axvr committed:
Update indentation options sections of the README and vim help file

axvr committed:
Improve some of the comments in the indentation code

axvr committed:
Move string indent picker to a separate function

axvr committed:
Fix false positive multi-line string detection

Sometimes it seems that when `=`ing over a block, Vim will sometimes
not re-highlight strings correctly until we have already ran
`searchpairpos`.  In this case, we implement a hacky workaround which
detects the false-positive and recovers from it.

axvr committed:
EDN files are unlikely to contain many lists, so check them last

axvr committed:
Yet more string detection fixes and slight code clean up

axvr committed:
Initial work on list indentation

Still lots to do here, this is just the initial work.

axvr committed:
More indentation code clean up and readability improvements

axvr committed:
Start of a small Clojure reader for indentation w/o syntax highlighting

Not complete yet, but getting there.

axvr commented:
I've been experimenting with a custom algorithm as per the check list item above.  This version eliminates all syntax highlight checks (and doesn't require any hacks to make those checks work) plus it is both faster and less code still.

Another benefit of this one is that since the bottle neck is no longer syntax highlight checks, by writing a version of the algorithm in Vim9 script, we can get it even faster still!  (Maintaining both a legacy Vim script and Vim9 script versions should be feasible due to how much simpler this is than the original implementation.)

TL;DR: I'll push a new version to this branch in a few days which will be:
1. Faster,
2. Have potential to become significantly faster (Vim9 script),
3. Simpler with no hacks,
4. Less code,
5. Works even when syntax is off.

Once this branch is feature complete, I'll share benchmarks comparing it to the original as is on `master`.

axvr committed:
Completed the mini Clojure reader as the core of the indentation system

Some refactoring should be possible here and further optimisations.
Once all optimisations I can think of have been implemented, I'll try
writing an alternate Vim9 script version.

(The syntax highlight group checks used in previous implementations of
the indentation code was the core bottleneck, so a Vim9 script version
would not have been much faster.)

axvr committed:
Reader indent algorithm perf has surpassed the prev syntax highlight one

The performance of the new reader-like indentation algorithm has now
surpassed the performance of my previous syntax highlighting approach
with better accuracy and no hacky code required.

axvr committed:
Small refactor to improve code clarity in indentation code

axvr commented:
Performance benchmarks as of current change.  This is measuring the time to reindent a whole file on Macbook Pro 2021, so expect much more noticable performance gains on slower hardware.

| Code version | 200 lines of Clojure | 600 lines of EDN |
|---|---|---|
| Original | 0.96s | 3.5s |
| New (legacy) | 0.33s | 2.6s |
| New (Vim9) | 0.06s | 0.46s |

(I'll share new benchmarks (and the full reports generated by Vim) once this branch is feature complete.)

With the Vim9 implementation in the code too (not pushed yet, still figuring out how best to integrate it in a maintainable way), we can expect the file size of `indent/clojure.vim` to be ~150 LOC less than the original.

axvr committed:
Fix indentation bug during comment detection

axvr committed:
Fix accidental detection of backslashes as tokens

Previously backslashes were accidentally detected as tokens by the
indentation tokeniser.  This meant that character literals, would break
indentation of everything after them.

axvr committed:
Update indent comments and move `s:Conf` function to the top

axvr committed:
Add basic function parameter alignment indentation

This likely still needs a lot of improvements.

axvr committed:
Replace `clojure_align_subforms` with `clojure_indent_style`

axvr committed:
Update and add a bunch of indentation tests

axvr committed:
Fix indentation for multibyte characters

axvr committed:
Switch test runner to Kaocha for nicer output

axvr committed:
Only run GitHub Actions workflow once for PRs

axvr committed:
Neovim does not have a `state()` function

axvr committed:
Initial work on building up macro indent rules

axvr committed:
Fix false-positive comment detection within strings

axvr committed:
Improve accuracy of multiline string detection for indentation

axvr committed:
Fix indentation when file contains multibyte characters

axvr committed:
Add more indent test cases

NoahTheDuke commented:
Any updates on this?

axvr commented:
@NoahTheDuke yes, slow steady progress during the limited time I have to work on it at the moment.  GitHub doesn't show it well, but there have been a bunch of bug fixes over the last few weeks.  Currently my focus in on getting the majority of the tests to pass.

If you wanted to use it, this branch is actually usable if you don't mind a few minor formatting issues.

NoahTheDuke commented:
Cool. I mistakenly trusted Github's "added 11 commits 2 months ago", which it seems bundles recent stuff too. I don't mean to pressure, just saw a bunch of activity then looked like nothing. Glad to hear you're still poking at it. I'll give it a whirl locally!

NoahTheDuke commented:
Just tried it out and it's great so far. Thanks so much!

axvr committed:
Neovim has added the `state()` function; use it if available

See: neovim/neovim#23576

Unintentional multiline string indentation will no longer occur in
future Neovim versions.

axvr committed:
Fix application of inline comments during indentation

Previously indenting code containing inline comments like this:

    {:foo 1  ; Hello [
     :bar 2}

Would have resulted in this incorrect indentation:

    {:foo 1  ; Hello [
                      :bar 2}

axvr committed:
Add tests to check that comments don't affect indentation

axvr committed:
Fix indentation of `with-*` macros

axvr committed:
Update comments and add `with-in-str` indent rule

axvr committed:
Improve accuracy of function operand indenting

In the original indent script, code like this:

    ((foo) 1
               3)

Would be indented as this:

    ((foo) 1
     3)

Now it is *correctly* indented as this:

    ((foo) 1
           3)

axvr commented:
Update: nearing readiness!  :tada:

This branch has now surpassed the original indentation code in pretty much every aspect (~2–3x faster, less and simpler code, more accurate indentation, better defaults, more customisable, not dependent on syntax highlighting, etc.).  There are a few small known bugs left to fix, the documentation needs updating and a little bit of code clean up will be required.

(The greatly expanded test suite is also already [paying dividends](https://idioms.thefreedictionary.com/pay+dividends); helping catch accidental regressions!  More tests to come soon.)

axvr committed:
Slight code refactor and added more comments

axvr committed:
When `lispoptions` feature is available, we can enable the `lisp` option

Closes: #40

axvr committed:
Indent some macros a bit like functions

axvr committed:
More indent test cases

axvr committed:
Fix escape character detection logic

axvr committed:
Simplify character column resolution

axvr committed:
Add some indent rules for more built in macros

SavageMessiah commented:
Thanks for working on this, I've been trying clojure dev in neovim lately and was getting really frustrated by the fact that it would indent based on brackets in string literals. I just tried this branch and that problem is fixed.

axvr committed:
Don't use fn arg alignment indent when a keyword is in function position

Fixes indentation of reader conditional macros.  May choose to remove
this if function argument indentation is generally preferred by
programmers when a keyword is in function position.

axvr committed:
Merge 'origin/master' into indent-forms

axvr commented:
> Thanks for working on this, I've been trying clojure dev in neovim lately and was getting really frustrated by the fact that it would indent based on brackets in string literals. I just tried this branch and that problem is fixed.

@SavageMessiah great to hear that!  If you encounter any issues on this branch please let me know.

axvr committed:
Add extra indentation test cases for records and protocols

NoahTheDuke commented:
Feedback:

```clojure
(filter-test
 a b c)

(filter-suite
  a b c)

(letfn [(filter-test
         [test-case]
         ...)
        (filter-suite
          [suite]
          ...)]
  ...)
```

It seems the matching algorithm for function indentation is based on any portion of the function's name, not the whole thing, so `test` is indented with 1 space and others are indented with 2. Regardless of my preference (1 vs 2), these should not be indented differently, especially when they're both user-defined functions.

Additionally, `letfn` functions have hanging indentation, not flow indentation, as seen below. The `func` body should be indented 2 spaces, not even with the brackets.

```clojure
(letfn [(func [test-case]
              ...)]
  ...)
```

Thanks so much!

axvr commented:
Hi @NoahTheDuke, thanks for raising and testing!

> It seems the matching algorithm for function indentation is based on any portion of the function's name, not the whole thing, so test is indented with 1 space and others are indented with 2. [...]

Unfortunately I'm not seeing the same behaviour you are with `filter-test` vs. `filter-suite`, perhaps you have some config options set controlling the indentation for those?  If you do and can share it, I'll happily take a further look into this issue.

> Additionally, letfn functions incorrectly have hanging indentation, not flow indentation [...]

I am aware of this current deficiency in the indentation algorithm.  It is possible to solve, but it will be tricky as `letfn` falls into what I call _complex indentation_, which includes others such as `extend-protocol`, `extend-type`, and method bodies of `defrecord`, `deftype` and `defprotocol`.  I call it "complex indentation" as they alter the indentation rules of the lists within them.

I will fix it, but it will likely be **after** this PR is merged as their use isn't particularly common and hopefully the other improvements I've made will make up for it being a temporary issue.  As a partial workaround, I'd recommend using `let` and `fn` instead of `letfn` _where possible_, as it's often no more code and is (at least in my opinion) more readable too.  Alternatively you can run `cljfmt` before committing or hook it into Vim with a pre-write autocommand.

---

### Asides

<details>
<summary>
Why is indenting <code>letfn</code> so difficult?  (Click here for more info.)
</summary>
<br>
<p>
The challenge with the complex indentation rules is in making it both efficient within Vim script, and generic/customisable.
</p>
<p>
Emacs' clojure-mode avoids the first (by Emacs Lisp being much, much faster than Vim script :sob:) and solves the latter nicely with a way of defining complex indentation rules (which I <a href="https://github.com/weavejester/cljfmt">cljfmt</a> largely replicated).  I may end up doing something similar, but do wonder if some flexibility could be sacrificed in favour of easier customisation and better performance...
</p>

<p>
(<code>letfn</code> is actually even worse than the other <i>complex indentation</i> rules as its rules are unique!  In the old clojure.vim indentation code we had <a href="https://github.com/clojure-vim/clojure.vim/blob/bc3f1faf93ac47343b66b0e189990149a7fc0715/indent/clojure.vim#L175-L199">hardcoded rules</a> just to handle it!)
</p>
</details>

<details>
<summary>
Click this to view a complete tangent on my opinions about <code>letfn</code> overuse (that no one asked for).  🙂
</summary>
<br>
<p>
Personally I feel that <code>letfn</code> is overused in Clojure code, as the only reason I can see to use it over plain <code>let</code> + <code>fn</code> is when the <code>letfn</code>'d functions need to call each other.  If it were used only in that case, it could be useful as a signal that that is happening, much like how use of <code>when</code> over <code>if</code> signifies that there is no "else" branch.
</p>
<p>
(The equivalent of <code>letfn</code> in Scheme is <code>letrec</code> — i.e. let recursive — which I believe may have inspired <code>letfn</code> and kind of backs up my thoughts on its overuse.)
</p>
</details>

axvr committed:
Begin refinements to configuration options

axvr committed:
Fix config option lookup

Vim script is infuriating sometimes!  Turns out that the indentation
tests locally passed with the previous code, but failed in CI and real
life.

axvr committed:
Discovered a way to put comments within dict defs in Vim script

axvr committed:
Improve indentation in "uniform" style and undo reader conditional work

When the "uniform" indent style is selected, we no longer apply other
list indentation rules/analysis.

Unfortunately the reader conditional changes broke indentation for the
following:

    (ns my-ns
      (:require [clojure.string :as str]
                [clojure.spec.alpha :as s]))

Indenting it as this:

    (ns my-ns
      (:require [clojure.string :as str]
       [clojure.spec.alpha :as s]))

Which is incorrect.

axvr committed:
Update README to better cover the new indentation options

axvr committed:
Replace the old multi-line string config option

axvr committed:
Remove the link to the #vim IRC as the link is now dead

axvr committed:
Fix typos and re-add the insert-mode completion info to the README

axvr committed:
Update the "About" section within the Vim help doc

axvr committed:
Update indentation section of the Vim help doc

axvr committed:
Improve indentation style examples in Vim help doc

axvr committed:
Minor documentation and comment refinements

axvr committed:
In README, mention the replacements for each old indent config option

axvr committed:
Add indentation tests covering keywords in function position

axvr committed:
Update indent regular expressions to set the required 'magic' mode

Makes indentation more robust against users toggling the `'magic'`
option.

axvr committed:
Compress comments on indent code

axvr committed:
Compress more comments on indent code

axvr committed:
Use `<kbd>` to show keyboard shortcuts in `README.md`

axvr committed:
Improved accuracy of first function argument detection in indentation

Indentation improvements:
    - Fixed escape character detection `s:IsEscaped`.
    - Improved accuracy of first function argument detection.

axvr committed:
Slightly improve keyboard shortcut formatting in `README.md`

axvr committed:
Set mode in new regexprs

axvr committed:
Make `ClojureIndent()` function public

axvr committed:
Indent `with-` macros and functions like cljfmt default

axvr committed:
Run indentation tests against Vim AND Neovim

Closes: #39

axvr committed:
Install latest Vim and Neovim in GitHub Actions

axvr committed:
Merge latest changes from 'origin/master'

axvr committed:
New indentation test runner

axvr committed:
Improve new test runner output and detect script errors

axvr committed:
Speed up indentation tests on Vim

axvr committed:
Move `indenttime` script to `dev` folder as `time-indent`

axvr committed:
Update GitHub Actions

axvr committed:
Kaocha not needed anymore, so removing dependency

axvr committed:
Bump copyright year

axvr committed:
Remove unneeded config from `clj/profile.clj`

axvr committed:
Run indentation tests in Vim's silent Ex-improved mode

Avoids screen flickering during test execution.

axvr committed:
Reduce indentation test runner noise and add colour output

axvr committed:
GitHub Actions "annotations" feature doesn't work with colour output

axvr committed:
Simplify indentation test runner output on GitHub Actions

garrett-hopper commented:
FYI, the `"uniform"` `indent_style` isn't quite following Tonsky's [Better Clojure formatting](https://tonsky.me/blog/clojurefmt/)

> Multi-line lists that start with a symbol are always indented with two spaces,
> Other multi-line lists, vectors, maps and sets are aligned with the first element (1 or 2 spaces).

Multi-line lists that start with a non-symbol are currently indented with 2 instead of 1 space.

E.g.

```clojure
(ns namespace
  (:require
   [clojure.set :as s]))
```

The final line is indented 1 space because `:require` is keyword and not a symbol.

```vim
if indent_style ==# 'uniform' | return base_indent + 1 | endif
```

Not sure the best way way for this to check if the first list item is a symbol or not. 🤔

I wrote this for #21 before I realized this rewrite was being done. It's a pattern that will match symbols, adapted from the tree-sitter grammar. (Though it may be easier to just special case non-symbols as first list item.)

```lua
local suffixonly = ":#'0-9\\/" -- not allowed as first character
local forbidden = {
  [[\[\]\n\r\t \\]], -- brackets, whitespace, slashes
  '(){}"@~^;`,', -- parens, braces, special characters
  suffixonly,
}
local prefix = string.format("[^%s]", table.concat(forbidden))
local suffix = string.format([[\(%s\|[%s]\)*]], prefix, suffixonly)
local symbol = prefix .. suffix
```

```
[^\[\]\n\r\t \\(){}"@~^;`,:#'0-9\/]\([^\[\]\n\r\t \\(){}"@~^;`,:#'0-9\/]\|[:#'0-9\/]\)*
```

axvr commented:
Thanks for raising this @garrett-hopper.  I'll add it to my list and try to fix it soon.

(Yet more mildly irritating rules in the common Clojure indentation styles :laughing: :sob:)

axvr commented:
Apologies to anyone waiting patiently for this to be merged.  I haven't managed to find much time to dedicate to this and each time I come back to it, it requires a lot of context reloading as the standard Clojure indentation styles are shockingly all very complicated (contrary to Clojure itself) and to implement them efficiently in Vim script is forever painful.

Currently what exists in this branch is IMO already far superior to what is currently on the `master` branch, so my current plan is:

1. fix a few of the more major outstanding issues,
2. document the `clojure_indent_rules` option with warnings that it may change in the future,
3. merge, 🥳 
4. open issues for the remaining known problems that I don't consider to be dealbreakers,
5. gather more feedback on other issues with wider use,
6. fix issues critical to the core algorithm,
7. later complete #35 and #32 once the core algorithm is perfected to avoid needing to implement the same fixes 3 times.

I hope to get round to the first 4 soon, but can't make any promises on when I will do so.

NoahTheDuke commented:
Pretty exciting! Thanks for shepherding this! I've been using this branch exclusively since you opened the PR and I'm quite pleased with it.

NoahTheDuke commented:
Bug report: defining protocol methods in `defrecord` under "standard" still indents the body to match the arg vector:

```clojure
(defrecord Example [a b c]
  MyProto
  (method1 [this d e f]
    (+ a b c d e f)))
```
becomes
```clojure
(defrecord Example [a b c]
  MyProto
  (method1 [this d e f]
           (+ a b c d e f)))
```

It looks like there's tests covering this, so I'm not sure what's wrong.
